pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/security/ApexCRUDViolationRule.java

```diff
@@ -277,14 +296,69 @@ private void checkForCRUD(final AbstractApexNode<?> node, final Object data, fin
         if (variable != null) {
             final String type = varToTypeMapping.get(Helper.getFQVariableName(variable));
             if (type != null) {
-                StringBuilder typeCheck = new StringBuilder().append(node.getNode().getDefiningType().getApexName()).append(":")
-                        .append(type);
+                StringBuilder typeCheck = new StringBuilder().append(node.getNode().getDefiningType().getApexName())
+                        .append(":").append(type);
 
                 validateCRUDCheckPresent(node, data, crudMethod, typeCheck.toString());
             }
         }
     }
 
+    private HashSet<ASTMethodCallExpression> getPreviousMethodCalls(final AbstractApexNode<?> self) {
+        final HashSet<ASTMethodCallExpression> innerMethodCalls = new HashSet<>();
+
+        final ASTBlockStatement blockStatement = self.getFirstParentOfType(ASTBlockStatement.class);
```

Comment:
checking up just one block is probably not enough. Think of something such as:

```
if (checkPerms()) {
  if (myOtherCondition) {
    // DML here
  } else {
    // DML here
  }
}
```

The blocks would just contain the DML, and not the check (assuming blocks are defined similarly to the Java grammar, since that's on imported jars I'm not certain).

You should probably recursively move up block by block until you reach the method block (avoid getting into the class block). This approach will also guarantee you are analyzing the correct branches only, and not looking into sibling blocks that are not in the current block's execution path.